crates/sui-indexer/src/handlers/pruner.rs

```diff
@@ -38,77 +41,268 @@ impl Pruner {
+                .get_min_cp_and_tx_for_epoch(new_inclusive_epoch_lower_bound)
+                .await?;
+            let new_lo = value.map_to_bound(cp, tx);
+            lower_bound_updates.push(Watermark::lower_bound(
+                *key,
+                new_inclusive_epoch_lower_bound,
+                new_lo,
+            ));
+        }
+    }
+
+    if !lower_bound_updates.is_empty() {
+        store.update_watermarks(lower_bound_updates).await?;
+        info!("Finished updating lower bounds for watermarks");
+    }
+
+    Ok(())
+}
+
+/// Task to periodically query the `watermarks` table and update the lower bounds for all watermarks
+/// if the entry exceeds epoch-level retention policy.
+async fn update_watermarks_lower_bounds_task(
+    store: PgIndexerStore,
+    epochs_to_keep: u64,
+    cancel: CancellationToken,
+) -> IndexerResult<()> {
+    loop {
+        if cancel.is_cancelled() {
+            info!("Pruner watermark lower bound update task cancelled.");
+            return Ok(());
+        }
+
+        update_watermarks_lower_bounds(&store, epochs_to_keep).await?;
+        tokio::time::sleep(Duration::from_secs(5)).await;
+    }
+}
+
+/// A task that queries `watermarks` table for the source of truth for the epoch range, and drops
+/// older partitions up to but not including `watermark.epoch_lo`.
+async fn prune_epoch_partitioned_tables_task(
+    store: PgIndexerStore,
+    partition_manager: PgPartitionManager,
+    cancel: CancellationToken,
+) -> IndexerResult<()> {
+    loop {
+        if cancel.is_cancelled() {
+            info!("Pruner prune_epoch_partitioned_tables_task cancelled.");
+            return Ok(());
+        }
+
+        let watermarks = store.get_watermarks().await?;
+
+        // Not all partitioned tables are epoch-partitioned, so we need to filter them out.
+        let table_partitions: HashMap<_, _> = partition_manager
+            .get_table_partitions()
+            .await?
+            .into_iter()
+            .filter(|(table_name, _)| {
+                partition_manager
+                    .get_strategy(table_name)
+                    .is_epoch_partitioned()
+            })
+            .collect();
+
+        // `watermarks` table is the source of truth for the epoch range. The partitions to drop are
+        // `[watermark.pruner_lo(), watermark.epoch_lo)`
+        for (table_name, (_, _)) in &table_partitions {
+            let Some(lookup) = WatermarkEntity::from_str(table_name) else {
+                // TODO: (wlmyng) handle this error
+                println!(
+                    "could not convert table name to WatermarkEntity: {}",
+                    table_name
+                );
+                continue;
+            };
+
+            let Some(entry) = watermarks.get(&lookup) else {
+                println!("coudl not find entity in watermarks: {:?}", lookup);
+                continue;
+            };
+
+            wait_for_prune_delay(&entry, &cancel, 1000).await?;
```

Comment:
Why is `wait_for_prune_delay` only called here and not in lookup table pruning?

Reply:
it'll be called in lookup table pruning, but just cleaning that part up rn so didn't get to it yet